[export] Better state_dict and constant dedup in torch.export.save

yiming0416 opened the pull request:
Summary:

Previously, weight deduplication was done by simply grouping tensors with their untyped storage and saving the first tensor in the group.

A more rigorous approach would be to find a complete tensor that covers the storage and store that tensor. This is particularly important for GPU weights because when saving to raw bytes, we move the weight to CPU first, and if the weight being saved is not a complete one, it will lose the storage information during the copy to CPU.

In this diff, we reuse code in `_package_weights.py` for better weights and constants deduplication in `torch.export.save`.

Test Plan: buck2 run mode/dev-nosan caffe2/test:test_export -- -r test_weight_sharing_gpu

Differential Revision: D83523690

yushangdi commented on test/export/test_serialize.py:
```diff
@@ -924,6 +925,27 @@ def forward(self, x):
+        save(ep, buffer)
+        buffer.seek(0)
+        loaded_ep = load(buffer)
+        self.assertEqual(m(*sample_inputs), loaded_ep.module()(*sample_inputs))
```
nit: I think this test would pass even if c1 and c2 are saved as two separate tensors? Might be good to also check that c1 is a slice of c2 in the loaded ep.

yiming0416 commented on torch/export/pt2_archive/_package_weights.py:
```diff
@@ -1,39 +1,58 @@
+        if tensor.is_contiguous():
+            return tensor.view(-1)[-1].data_ptr() + tensor.element_size()
+        else:
+            stop = tensor.reshape(-1)[-1].data_ptr() + tensor.element_size()
```
@yushangdi  good catch! now always return False in `is_complete()` when the tensor is not contiguous

yiming0416 committed:
[export] Better state_dict and constant dedup in torch.export.save (#164196)

Summary:

Previously, weight deduplication was done by simply grouping tensors with their untyped storage and saving the first tensor in the group.

A more rigorous approach would be to find a complete tensor that covers the storage and store that tensor. This is particularly important for GPU weights because when saving to raw bytes, we move the weight to CPU first, and if the weight being saved is not a complete one, it will lose the storage information during the copy to CPU.

In this diff, we reuse code in `_package_weights.py` for better weights and constants deduplication in `torch.export.save`.

Test Plan: buck2 run mode/dev-nosan caffe2/test:test_export -- -r test_weight_sharing_gpu

Reviewed By: angelayi

Differential Revision: D83523690
